app/src/main/java/de/hhn/aib/labsw/blackmirror/controller/PageController.java

```diff
@@ -194,4 +202,74 @@ private void autoChangePageByTime() {
             }
         }
     }
+
+    @Override
+    public void dataReceived(String topic, JsonNode object) {
+        super.dataReceived(topic, object);
+
+        assert (topic.equals(PAGE_UPDATE_TOPIC)) : "Wrong topic received by PageController.";
+
+        //reset the current pages
+        for (int i = 0; i < pages.size(); i++) {
+            goToAnyPage(i);
+            getCurrentPage().setWidgetsInvisible();
+            deletePageAtIndex(i);
+        }
+
+        try {
+            LayoutData data = nodeToObject(object, LayoutData.class);
+
+            // iterate the pages
+            for (int i = 0; i < data.pages().size(); i++) {
+
+                //create new page
+                ArrayList<AbstractWidgetController> page = new ArrayList();
+
+                // process all widgets for this page
+                if (data.pages().get(i).widgets().size() > 0) {
+                    for (int j = 0; j < data.pages().get(i).widgets().size(); j++) {
+                        WidgetData widget = data.pages().get(i).widgets().get(j);
+                        switch (widget.type()) {
+                            case CALENDAR -> {
+                                CalendarWidgetController calendarWidgetController = new CalendarWidgetController();
+                                calendarWidgetController.getWidget().setPosition(widget.x(), widget.y());
+                                page.add(calendarWidgetController);
+                            }
+                            case CLOCK -> {
+                                // todo : FaceType handling
```

Comment:
This could be done by seperately requesting (by publishing a "requestXYZ" message) the clock face type.
Alternatively we could send the custom widget config data with the page layout data in the first place, although this would require quite a view changes both in the mirror and in the app code

Reply:
I suggest we handle this case in the next sprint, since the changing of the FaceType is low priority anyway.
I just set the FaceType.ANALOG as default for now.